### PR TITLE
V2 §2: Path identity — drop dot-joined strings

### LINKED ISSUE #246: V2 §2: Path identity — drop dot-joined strings
Part of the [V2.0 roadmap](V2-roadmap.md). Foundational identity model.

## Problem

[`toPathString`](src/helpers.ts) joins keys with `.` and patches empty strings with `\0`. Keys containing dots produce ambiguous paths. Worse, `areChildrenBeingEdited` in [TreeStateProvider.tsx](src/contexts/TreeStateProvider.tsx) does `pathString.includes(...)` — a **substring** match, so `"foo"` claims `"foobar"`'s children are being edited.

## Approach

Use `CollectionKey[]` as node identity everywhere; only stringify for React keys (and use a safe encoding there).

## Why this comes early

- The substring-matching fix in `areChildrenBeingEdited` is bundled with the §4 TreeStateProvider refactor (item 4 of that breakdown — proper prefix check on arrays).
- §16 Fine-grained re-rendering depends on stable node-identity-by-path keys.

### PR DESCRIPTION
**Pull request by CarlosNZ**
Closes #246. Part of the [V2.0 roadmap](../blob/v2.0-dev/V2-roadmap.md), §2.

## Summary

- Internal node identity is now `CollectionKey[]` everywhere instead of a dot-joined string.
- `areChildrenBeingEdited`, the drag-onto-self guard, and the editing-state comparisons all use array predicates (`pathsEqual`, `isDescendantOf`) — no more string `.includes()` / `.startsWith()` substring checks.
- `toPathString` survives as a public utility but its encoding changes to `/` + `encodeURIComponent` (injective by construction). The `key?: 'key_'` second arg is removed — mode is now a field on the editing-state object, not a string prefix.

## Bugs fixed

This kills three related bugs that were all rooted in the same string-encoding shortcut. Paste this into the demo to see them on `v2.0-dev` (and watch them vanish here):

\`\`\`json
{
  \"foo\": { \"name\": \"A\" },
  \"foobar\": { \"name\": \"B\" },
  \"with.dot\": \"I collide with 'with > dot'\",
  \"with\": { \"dot\": \"I collide with 'with.dot'\" }
}
\`\`\`

1. **Dotted-key collision.** Editing the root \`\"with.dot\"\` value used to put _both_ it and the nested \`with > dot\` value into edit mode simultaneously — they serialised to the same path string.
2. **Substring match in \`areChildrenBeingEdited\`.** Editing \`foobar > name\` would force the unrelated \`foo\` collection to open (because \`\"foobar.name\".includes(\"foo\")\`).
3. **Substring match in drag-onto-self guard.** Dragging \`foo\` over \`foobar\` would hide the drop indicator (\`\"foobar\".startsWith(\"foo\")\`).

Also fixes a related latent bug in \`handleDrop\` ([useDragNDrop.tsx:122–123](../blob/v2.0-dev/src/hooks/useDragNDrop.tsx#L122)) where \`sourceBase\` was joined with \`.\` but \`thisBase\` was joined with \`''\` — so \`KEY_EXISTS\` fired spuriously on legitimate same-parent drags at depth ≥ 3.

## Implementation shape

- Editing state is now \`{ path: CollectionKey[]; mode: 'value' | 'key' } | null\`, bundled so transitions are atomic (the §4 reducer work in #247 can grow this into a discriminated union naturally).
- \`previouslyEditedElement\` stays path-only (tab nav only ever targets value-mode).
- \`dragSource\` drops the mirrored \`pathString\` field.
- New helpers in \`src/helpers.ts\`: \`pathsEqual\`, \`isDescendantOf\` (reflexive — a path is a trivial descendant of itself; both callers want that), \`editingStatesEqual\`.

## Bundle size impact

Essentially a wash — adding new symbol names slightly reduces gzip efficiency:

| | Raw | Gzipped |
|---|---|---|
| ESM | 53,421 → 53,418 (−3 bytes) | 18,054 → 18,072 (+18 bytes, +0.10%) |
| CJS | 54,589 → 54,587 (−2 bytes) | 18,104 → 18,115 (+11 bytes, +0.06%) |

## Test plan

- [x] New unit tests in \`test/pathHelpers.test.ts\` cover \`pathsEqual\`, \`isDescendantOf\` (including reflexive case + substring-lookalike regression), and \`toPathString\` (encoding injectivity, slashes, percents, empties, unicode).
- [x] \`pnpm lint && pnpm compile && pnpm test\` — all green (126 tests).
- [x] \`pnpm -r build\` — core + themes + components all build cleanly.
- [x] Demo verification: paste the JSON above, confirm all three bugs are gone.
- [x] Tab navigation through the tree (exercises \`previouslyEditedElement\`).
- [ ] External \`edit\` trigger fires correctly on the matching path.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/helpers.ts [hunk 1]
```diff
@@ -126,15 +127,32 @@ export const matchNodeKey: SearchFilterFunction = ({ key, path }, searchText = '
...
  */
-export const toPathString = (path: Array<string | number>, key?: 'key_') =>
-  (key ?? '') +
-  path
-    // An empty string in a part will "disappear", so replace it with a
-    // non-printable char
-    .map((part) => (part === '' ? String.fromCharCode(0) : part))
-    .join('.')
+export const toPathString = (path: CollectionKey[]) =>
+  path.map((part) => encodeURIComponent(String(part))).join('/')
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/helpers.ts]: Good catch — the injectivity claim was overstated. Fixed in f25c49b:

- `['']` now maps to `'\0'` via a single-case short-circuit before the `join`-based encoding runs. Safe because `encodeURIComponent` never emits a literal null char (it produces `'%00'` for the null byte).
- All other empty-key positions already disambiguate via separator structure, so they're unchanged: `['a', '']` → `'a/'`, `['', 'a']` → `'/a'`, `['', '']` → `'/'`, `['', '', '']` → `'//'`, `['a', '', 'b']` → `'a//b'`.
- Added explicit tests for each empty-key position plus a covering injectivity check over the edge-case paths (including `['\0']` to verify the sentinel doesn't collide with a literal null-byte key).

Docstring updated to document the sentinel rather than overstate the guarantee.
